File: java/add_and_search_word_data_structure_design.java
public class WordDictionary {

    class TrieNode {
        boolean isWord;
        Map<Character, TrieNode> children;
    }
    
    TrieNode root;
    
    WordDictionary() {
        root=new TrieNode();        
    }
    
    // Adds a word into the data structure.
    public void addWord(String word) {
        int i;
        TrieNode node=root;
        char[] wordArray=word.toCharArray();
        for (i=0; i<wordArray.length; i++) {
            if (node.children==null || !node.children.containsKey(wordArray[i])) break;
            node=node.children.get(wordArray[i]);
        }
        
        for (; i<wordArray.length; i++) {
            TrieNode curr=new TrieNode();
            if (node.children==null) {
                node.children=new TreeMap<Character, TrieNode>();
            }
            node.children.put(wordArray[i], curr);
            node=curr;
        }
        node.isWord=true;
    }

    boolean search(char[] word, int start, TrieNode node) {
        if (start==word.length) {
            return node.isWord;
        }    
        
        if (word[start]!='.') {
            if (node.children!=null && node.children.containsKey(word[start])) {
                node=node.children.get(word[start]);
                return search(word, start+1, node);
            } else {
                return false;
            }
        } else {
            if (node.children!=null) {
                for (Map.Entry<Character, TrieNode> entry:node.children.entrySet()) {
                    if (search(word, start+1, entry.getValue())) return true;
                }
            }
            return false;
        }
    }
    
    // Returns if the word is in the data structure. A word could
    // contain the dot character '.' to represent any one letter.
    public boolean search(String word) {
        return search(word.toCharArray(), 0, root);
    }
}

// Your WordDictionary object will be instantiated and called as such:
// WordDictionary wordDictionary = new WordDictionary();
// wordDictionary.addWord("word");
// wordDictionary.search("pattern");

File: python/add-and-search-word-data-structure-design.py
class TrieNode:
    def __init__(self):
        self.children=collections.defaultdict(TrieNode)
        self.isWord=False
        
class WordDictionary:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root=TrieNode()

    def addWord(self, word: str) -> None:
        """
        Adds a word into the data structure.
        """
        node=self.root
        for w in word:
            node=node.children[w]
        node.isWord=True

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        """
        def dfs(word, start, node):
            if start==len(word): return node.isWord
            if word[start]=='.':
                for v in node.children.values():
                    if dfs(word, start+1, v): return True
                return False
            else:
                if word[start] not in node.children: return False
                return dfs(word, start+1, node.children[word[start]])
            
        return dfs(word, 0, self.root)        


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

